### PR TITLE
VZ-4415 Create acceptance tests for overriding the default metrics templates

### PR DESCRIPTION
**Pull request by davisgomes**
# Description

This PR includes acceptance tests that include overriding the default metrics templates. There are three tests:

1. A namespace specifying a specific template
2. A workload specifying a specific template
3. A non-explicit but matching template in the workload namespace that takes precedence over the default template

Additionally, there are updates to the binding generator webhook that allow users to annotate the namespace to specify templates.

Fixes VZ-4415

# Checklist 

As the author of this PR, I have:

- [ ] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: application-operator/controllers/webhooks/metrics-binding-generator-workload.go [hunk 1]
```diff
@@ -76,16 +77,32 @@ func (a *GeneratorWorkloadWebhook) handleWorkloadResource(ctx context.Context, r
...
 	if metricsTemplateAnnotation, ok := unst.GetAnnotations()[MetricsAnnotation]; ok {
 		if metricsTemplateAnnotation == "none" {
 			log.Debugf("%s is set to none - opting out of metrics", MetricsAnnotation)
 			return admission.Allowed(constants.StatusReasonSuccess)
 		}
 	}
 
+	// If "none" is specified for annotation "app.verrazzano.io/metrics" then this namespace has opted out of metrics.
+	if metricsTemplateAnnotation, ok := workloadNamespace.GetAnnotations()[MetricsAnnotation]; ok {
+		if metricsTemplateAnnotation == "none" {
```
Comment (kminder) [inline application-operator/controllers/webhooks/metrics-binding-generator-workload.go:98]: I'm wondering if this should be case insensitive?

Comment (davisgomes) [reply application-operator/controllers/webhooks/metrics-binding-generator-workload.go:98]: I don't see why not.

File: application-operator/controllers/webhooks/metrics-binding-generator-workload.go [hunk 2]
```diff
@@ -138,35 +155,40 @@ func (a *GeneratorWorkloadWebhook) handleWorkloadResource(ctx context.Context, r
...
+	namespacedName := types.NamespacedName{Namespace: unst.GetNamespace(), Name: metricsTemplate}
+	err := a.Client.Get(context.TODO(), namespacedName, template)
+	if err != nil {
+		// If we don't find the metrics template in the namespace of the workload resource then
+		// look in the verrazzano-system namespace
+		if apierrors.IsNotFound(err) {
+			namespacedName := types.NamespacedName{Namespace: constants.VerrazzanoSystemNamespace, Name: metricsTemplate}
+			err := a.Client.Get(context.TODO(), namespacedName, template)
+			if err != nil {
+				log.Errorw(fmt.Sprintf("Failed getting metrics template: %v", err), "Namespace", constants.VerrazzanoSystemNamespace, "Name", metricsTemplate)
```
Comment (kminder) [inline application-operator/controllers/webhooks/metrics-binding-generator-workload.go:179]: Should these not be using the Errorw now?

File: application-operator/controllers/webhooks/metrics-binding-generator-workload.go [hunk 3]
```diff
@@ -138,35 +155,40 @@ func (a *GeneratorWorkloadWebhook) handleWorkloadResource(ctx context.Context, r
...
+				return nil, err
+			}
+			log.Debugw("Found matching metrics template", "Namespace", constants.VerrazzanoSystemNamespace, "Name", metricsTemplate)
+			return template, nil
 		}
 
-		log.Debugw("Found matching metrics template", "Namespace", unst.GetNamespace(), "Name", metricsTemplate)
-		return template, nil
+		log.Errorw(fmt.Sprintf("Failed getting metrics template: %v", err), "Namespace", unst.GetNamespace(), "Name", metricsTemplate)
```
Comment (kminder) [inline application-operator/controllers/webhooks/metrics-binding-generator-workload.go:186]: Errorw?

File: application-operator/controllers/webhooks/metrics-binding-generator-workload.go [hunk 4]
```diff
@@ -138,35 +155,40 @@ func (a *GeneratorWorkloadWebhook) handleWorkloadResource(ctx context.Context, r
...
+		return nil, err
 	}
 
-	return nil, nil
+	log.Debugw("Found matching metrics template", "Namespace", unst.GetNamespace(), "Name", metricsTemplate)
```
Comment (kminder) [inline application-operator/controllers/webhooks/metrics-binding-generator-workload.go:190]: Debugw?

File: tests/e2e/metricsbinding/helidon-namespace-annotation/helidon_namespace_annotation_test.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+var t = framework.NewTestFramework("helidonnamespaceannotation")
+
+var _ = t.BeforeSuite(func() {
+	start := time.Now()
+	metricsbinding.DeployApplicationAndTemplate(namespace, yamlPath, templatePath, map[string]string{"app.verrazzano.io/metrics": "standard-k8s-metrics-template"})
+	metrics.Emit(t.Metrics.With("deployment_elapsed_time", time.Since(start).Milliseconds()))
+})
+
+var clusterDump = pkg.NewClusterDumpWrapper()
+var _ = clusterDump.AfterEach(func() {}) // Dump cluster if spec fails
```
Comment (kminder) [inline tests/e2e/metricsbinding/helidon-namespace-annotation/helidon_namespace_annotation_test.go:36]: If the comment is accurate this seems excessive.

File: application-operator/controllers/webhooks/metrics-binding-generator-workload.go (hunk 2 above)
Comment (davisgomes) [reply application-operator/controllers/webhooks/metrics-binding-generator-workload.go:179]: Looking at Bob's PR, it seems I'll have to reconcile some of the log messages

**Review by bobdonat — APPROVED** [review]

**Pull request merged**